src/AzerothWarsCSharp.Source/Researches/Ironforge/DeeprunTram.cs

```diff
@@ -1,62 +1,66 @@
-//       SetUnitInvulnerable(stormwindTram, false);
-//     }
-//
-//     private static void ResearchStart( ){
-//       var i = 0;
-//       while(true){
-//         if ( i > MAX_PLAYERS){ break; }
-//         Person.ById(i).SetObjectLimit(RESEARCH_ID, 0);
-//         i = i + 1;
-//       }
-//     }
-//
-//     private static void ResearchCancel( ){
-//       var i = 0;
-//       while(true){
-//         if ( i > MAX_PLAYERS){ break; }
-//         Person.ById(i).SetObjectLimit(RESEARCH_ID, 1);
-//         i = i + 1;
-//       }
-//     }
-//
-//     public static void Setup( ){
-//       RegisterResearchFinishedAction(RESEARCH_ID,  Transfer);
-//       RegisterResearchStartedAction(RESEARCH_ID,  ResearchStart);
-//       RegisterResearchCancelledAction(RESEARCH_ID,  ResearchCancel);
-//     }
-//
-//   }
-// }
+
+using static War3Api.Common;
+using AzerothWarsCSharp.Source.Setup.FactionSetup;
+using AzerothWarsCSharp.MacroTools.FactionSystem;
+using WCSharp.Events;
+using AzerothWarsCSharp.MacroTools.Libraries;
+using AzerothWarsCSharp.MacroTools;
+using WCSharp.Shared.Data;
+using AzerothWarsCSharp.Source.Setup.Legends;
+
+namespace AzerothWarsCSharp.Source.Researches.Ironforge
+{
+   public static class DeeprunTram
+   {
+
+   
+     private const int ResearchId = Constants.UPGRADE_R014_DEEPRUN_TRAM_IRONFORGE;
+   
+
+     private static void Transfer()
+     {
+       
+       
+       var Greatforge = LegendIronforge.LegendGreatforge.Unit;
+       var ironForgeLocation = new Point(GetUnitX(Greatforge), GetUnitY(Greatforge));
+       var Keep = LegendStormwind.LegendStormwindkeep.Unit;
+       var StormwindLocation = new Point(GetUnitX(Keep),GetUnitY(Keep));
+       unit tramToIronforge = PreplacedUnitSystem.GetUnit(FourCC("n03B"), StormwindLocation);
+       unit tramToStormwind = PreplacedUnitSystem.GetUnit(FourCC("n03B"), ironForgeLocation);
+       var recipient = IronforgeSetup.Ironforge?.Player ?? StormwindSetup.Stormwind?.Player;
+       if (recipient == null){
+         KillUnit(tramToIronforge);
+         KillUnit(tramToStormwind);
+         return;
+       }
+
+       SetUnitOwner(tramToIronforge, recipient, true);
+       WaygateActivate(tramToIronforge,true);
+       tramToIronforge.SetWaygateDestination(Regions.Stormwind.Center);
+       SetUnitInvulnerable(tramToIronforge, false);
+
+       SetUnitOwner(tramToStormwind, recipient, true);
+       WaygateActivate(tramToStormwind, true);
+       tramToStormwind.SetWaygateDestination(Regions.Ironforge.Center);
+       SetUnitInvulnerable(tramToStormwind, false);
+     }
+
+     private static void ResearchStart( ){
+        foreach (var player in GeneralHelpers.GetAllPlayers()){
+          player.GetFaction()?.ModObjectLimit(ResearchId, -1);
+        }
+     }
+
+      private static void ResearchCancel( ){
+        foreach (var player in GeneralHelpers.GetAllPlayers()){
+          player.GetFaction()?.ModObjectLimit(ResearchId, 0);
```

Comment:
I think this probably needs to be 1 rather than 0. Basically if someone cancels the research, it re-enables the research for everyone.